worker/src/websocket.rs

```diff
@@ -0,0 +1,267 @@
+    pub fn on_close<F: Fn(CloseEvent) + 'static>(&self, fun: F) -> Result<()> {
+        self.add_event_handler("close", move |event: worker_sys::CloseEvent| {
+            fun(event.into());
+        })
+    }
+
+    /// Registers an async event-handler for the `close` event.
+    ///
+    /// This event will be fired when the `WebSocket` closes.
+    ///
+    /// https://developers.cloudflare.com/workers/runtime-apis/websockets#events
+    pub fn on_close_async<
+        Fut: Future<Output = ()> + 'static,
+        F: Fn(CloseEvent) -> Fut + 'static,
+    >(
+        &self,
+        fun: F,
+    ) -> Result<()> {
+        self.add_event_handler("close", move |event: worker_sys::CloseEvent| {
+            let fut = fun(event.into());
+            wasm_bindgen_futures::spawn_local(async move { fut.await });
+        })
+    }
+
+    /// Registers an event-handler for the `error` event.
+    ///
+    /// This event will be fired when the `WebSocket` caught an error.
+    ///
+    /// https://developers.cloudflare.com/workers/runtime-apis/websockets#events
+    pub fn on_error<F: Fn(ErrorEvent) + 'static>(&self, fun: F) -> Result<()> {
+        self.add_event_handler("error", move |event: worker_sys::ErrorEvent| {
+            fun(event.into());
+        })
+    }
+
+    /// Registers an async event-handler for the `error` event.
+    ///
+    /// This event will be fired when the `WebSocket` caught an error.
+    ///
+    /// https://developers.cloudflare.com/workers/runtime-apis/websockets#events
+    pub fn on_error_async<
+        Fut: Future<Output = ()> + 'static,
+        F: Fn(ErrorEvent) -> Fut + 'static,
+    >(
+        &self,
+        fun: F,
+    ) -> Result<()> {
+        self.add_event_handler("error", move |event: worker_sys::ErrorEvent| {
+            let fut = fun(event.into());
+            wasm_bindgen_futures::spawn_local(async move { fut.await });
+        })
+    }
+
+    /// Internal utility method to avoid verbose code.
+    /// This method registers a closure in the underlying JS environment, which calls back
+    /// into the Rust/wasm environment.
+    /// Since this is a 'long living closure', we need to actively "leak" memory in this,
+    /// as we need to drop the reference to the `WasmClosure` or it will be destroyed with
+    /// it's lifetime at the end of the function.
+    fn add_event_handler<T: FromWasmAbi + 'static, F: FnMut(T) + 'static>(
+        &self,
+        r#type: &str,
+        fun: F,
+    ) -> Result<()> {
+        let js_callback = Closure::wrap(Box::new(fun) as Box<dyn FnMut(T)>);
+        let result = self
+            .socket
+            .add_event_listener(
+                JsValue::from_str(r#type),
+                Some(js_callback.as_ref().unchecked_ref()),
+            )
+            .map_err(Error::from);
+        // we need to leak this closure, or it will be invalidated at the end of the function.
+        js_callback.forget();
```

Comment:
Instead of leaking this, could we possibly just clean it up when we close the websocket? I haven't entirely wrapped my brain around what wasm-bindgen would think of that lifetime wise so if this isn't feasible just mark as resolved.

Reply:
I'm not 100% sure either. Without this call, the callback will not work in 100% of the cases.
Eventually you can hold that value somewhere in the depths of the code and release it once the websocket closes, but how do you have a guarantee to get that event?
1) It's the user's responsibility to call accept();
2) We'd need to always register an extra closure to catch that event internally as well.

All together, I think, the Closure is light enough, it doesn't really matter if it's static memory.
But again, I don't know how wasm keeps the rust program open or how it would evict that.

What would be your approach to caching the closures until... "something" happens?

Reply:
What *could* work, is we keep a vec of owned closures in a reference counted refcell inside of the websocket itself and then drop them all once we close. I just prototyped this out myself and found it's rather clunky and would remove the `PartialEq` and `Eq` trait impls. I'll mark this one as resolved for now and I can come back later if this becomes a problem.